### ISSUE TITLE
[idea] more compact way to indicate an account stores a particular struct?

### ISSUE THREAD

**Issue by mikemaccana** [coral-xyz/anchor#3184]
So right now we tell Anchor that space is derived from the (currently constant) discriminator size plus INIT_SPACE for the struct:

```
#[account(
    ...
    space = ANCHOR_DISCRIMINATOR + Offer::INIT_SPACE,
    ...
)]
pub offer: Account<'info, Offer>,
```

And as you mentioned in #3005 we could have just SPACE (including whatever size the discriminator is):

```
#[account(
    ...
    space = Offer::SPACE,
    ...
)]
pub offer: Account<'info, Offer>,
```

Is there any reason we couldn't do something like:

```
#[account(
    ...
    stores = Offer,
    ...
)]
pub offer: Account<'info, Offer>,
```

? The programmer's intention is to store a struct of a particular type. The fact that the account needs to match the size of the struct is really a dependency of that goal. Also it's easier to reason about!

**Comment by acheroncrypto** [comment]
Yeah, I've been working on something related that also requires automatic space calculation, so I'm thinking about making the `space` constraint optional again (it was like that a long time ago).

**Comment by jamie-osec** [comment]
This is fixed in v2, as the space is automatically emitted for any account types implementing it. `space` is still required for e.g. `UncheckedAccount` and will emit a compile error if not provided
